feat: add event dataclasses

cademirch opened the pull request:
This adds logevent dataclasses for conviently parsing log records into a dataclass.

This could also be extended to provide formatting to simplify logging upstream. This would move the complex formatting out of Snakemake's current DefaultFormatter and to the LogEvent. 

Example:

```python
#snakemake/src/snakemake/logging/events.py
from snakemake_interface_logger_plugins.common import JobInfo as BaseJobInfo

class JobInfo(BaseJobInfo):
    
    def format(self) -> str:
        lines = [f"rule {self.rule_name}:"]
        lines.append(f"    jobid: {self.jobid}")
        lines.append(f"    threads: {self.threads}")
        # ... custom formatting logic ...
        return "\n".join(lines)
```

```python
#snakemake/src/snakemake/jobs.py
...
def log_info(self, indent=False, printshellcmd=True):
    from snakemake.logging.events import JobInfo, ShellCmd
    
    priority = self.priority
    benchmark = (
        fmt_iofile(self.benchmark, as_input=False, as_output=True)
        if self.benchmark is not None
        else None
    )
    
    # Create event
    job_event = JobInfo(
        jobid=self.dag.jobid(self),
        rule_name=self.rule.name,
        rule_msg=self.message,
        # ... all the fields ...
        indent=indent,
    )
    
    # Format and log - the message is now a plain string
    formatted_msg = job_event.format(quiet=self.dag.workflow.output_settings.quiet)
    logger.info(
    formatted_msg, 
    extra={
        'event': LogEvent.JOB_INFO,
        'quietness': Quietness.RULES,
        **asdict(job_event),  # Unpacks all dataclass fields
    }
)
    
    # Shell command
    if printshellcmd and self.shellcmd:
        logger.info(self.shellcmd, extra={'event': LogEvent.SHELLCMD, 'quietness': Quietness.RULES, 'indent': indent})
    
    if self.rule.is_checkpoint:
        logger.info("DAG of jobs will be updated after completion.")
    if self.rule.is_handover:
        logger.info("Handing over execution to foreign system...")
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added typed dataclass representations for logging events and a runtime mapping to reconstruct event objects from log records, enabling structured, type-safe event access.

* **Documentation**
  * README updated with the dataclass-based event model, a consolidated event table, and usage examples for parsing and handling common event types (JOB_ERROR, JOB_FINISHED, PROGRESS).

* **Tests**
  * New test suite validating event conversion, alias handling, mapping consistency, and run-info/stat shapes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

cademirch committed:
feat: add event dataclasses

cademirch committed:
docs: update docs for event dataclasses

johanneskoester reviewed (approved):
I like this a lot!! Very nice design.

jlumpe commented:
- I'd say it's probably not as important to add the complete set of new events to this PR as those are easy to add afterward without interfering with any of the existing stuff.
- I feel like some attributes have default values applied in `from_record()` when they shouldn't - I'm not sure it makes sense to set the job ID to 0 instead of just raising an exception when not found.
- This is somewhat of a nitpick, but I've noticed that some of the attributes are inconsistently named (`jobid` vs `job_id` and `groupid` vs `group_id`). If you wanted to change those, this is probably the best time to do it. You would also not have to break backwards compatibility - the `from_record()` method (and I'm also envisioning an inverse like `to_extra()` to get the `extra` argument to pass to logging functions) could also include the old name of the attribute.  I think this is already being done with `JOB_STARTED`, `jobs` on the record corresponds to `job_ids` on the dataclass.
- Also a nitpick, but it might be simpler for the collection-typed fields of `JobInfo` (like `input` / `output` or `wildcards`) to default to an empty collection instead of `None`.

I've also made some of my own additions in #45 for reference.

jlumpe commented:
I'm also somewhat confused about the `ERROR` and `JOB_ERROR` events (my original motivation for contributing to the logging system was for improving the error reporting). As far as I can tell Snakemake is not currently setting any of the attributes for the `ERROR` event other than `exception`, which is set to the name of the exception class (the only use of the event I could find is in `snakemake.exceptions.print_exception()`)). If it's for logging generic exceptions, Python's logging system already has a standard way to include the exception type and instance (and traceback) through `LogRecord.exc_info`.

I feel like the error-related events should only include the Snakemake-specific information. Here's a list of what I think that could be:

- For Python exceptions that occur in a user-defined workflow file:
    - If it was a Snakefile or script file.
    - Absolute path path to it (may not be the main `Snakefile` if an `include` statement was used).
    - Workflow-relative path (not always applicable because you can run scripts from anywhere).
    - Mapping between line numbers in the original source file and those in the associated traceback object (which is for some transformed version of the original source). The  `print_exception()` function I mentioned already takes this as an argument.
 - General information on the runtime context:
   - Name of associated rule, if any.
   - ID of running job, if any.

This could all be implemented in one or multiple event types.

jlumpe committed:
Updates to event dataclasses (#45)

- Added `LogEventData` base class.
- Added explicit mapping between `LogEvent` enum values and
`LogEventData` subclasses (`LogEventData.event` class attribute, and
`LOG_EVENT_CLASSES` global var).
- Added default implementation of `from_record()` class method that just
inspects the dataclass fields:
- Allows attributes to be missing from the record if the corresponding
field has a default value, raises an exception otherwise.
  - Most subclasses just use the default implementation.
- This changes the previous behavior in a few places, where a default
value was applied in `from_record()` that was different than the
dataclass field default.

Still to be done:

- Add `data_from_record()` function which dispatches to proper subclass
based on `record.event`.
- Should the `from_record()` class method be made private in this case?
- Add `to_extra()` method, which converts to the dict to be passed as
the `extra` argument to a logging function.
  - Tests for round trip of data to/from `LogRecord` instance.

cademirch committed:
Merge branch 'main' into feat-event-dataclasses

cademirch committed:
dpes: pin packaging to pyproject.toml

cademirch committed:
Merge branch 'main' into feat-event-dataclasses

cademirch committed:
deps: remove git interface-common due to conflict with packagaing

cademirch committed:
make type checker happy

cademirch committed:
linting

cademirch committed:
add aliases; check canonical name first

jlumpe commented:
Interested in hearing more about it, is there another developer meeting coming up?

cademirch commented:
> Interested in hearing more about it, is there another developer meeting coming up?

I'll post an issue with an rfc and tag you soon! next meeting will be in april, skipped this month since hackathon
